### DISCUSSION TITLE
sfdx-git-delta picking up previous delta changes

### DISCUSSION THREAD

**Discussion by j-obrien97** [scolladon/sfdx-git-delta#674] [Q&A] [answered]
Experiencing an issue where the delta package is created successfully, however, when going to make another change in the repository, on the pr build (and completion) pipeline included the previous delta changes.

Pipeline is an Azure DevOps Pipeline running on a YAML file. 

Main guts of the YAML below: 
 
- script: |
     sfdx update
  displayName: Install SF CLI
- script: echo y | sfdx plugins:install sfdx-git-delta
  displayName: Install SFDX Git Delta
- script: |
    git config remote.origin.fetch '+refs/heads/*:refs/remotes/origin/*'
    git fetch --all
    sfdx sgd:source:delta -f origin/master --output .
  displayName: Creating Delta
- script: sfdx force:source:convert --manifest package/package.xml --output-dir=convert 
  displayName: Convert to MDAPI

I would expect that after the PR completes, 'origin/master' would include the previous changes. Am i incorrect?

**Comment by scolladon** [comment]
Hi @j-obrien97 !

Thanks for those information.

I have questions:
- Does the changes are in the same PR ?
- Does the next PR is from another branch ?
- Could you explain with a bit more detail how it works ? I would like to reproduce it myself so I can be sure to completely understand the situation ?

**Reply by j-obrien97** [reply]
Hi @scolladon absolutley!

The repo im working in only has one branch 'master'

So the process is as follows.
(initial change)
1- Using VS Code, pull down the master branch, create feature branch 'change1', make changes (update a field label etc..)
2- Push code up into PR
3- PR runs the build validation running through the pipeline steps from my initial question
4- Complete PR, Build pipeline runs again, creates Delta package successfully
(second change)
Run though steps 1 & 2 to push another change to the repo
Now when the build pipeline runs (on pr creation AND completion) it will still produce a 'new' delta. However, that delta contains the previous change, which i can confirm was successfully merged into the master branch before the new changes were submitted.

So if i were to make a change to the an Account fields label and run through the whole Delta process. Complete that PR. On the next PR to the same repo/branch, the pipeline will produce a delta containing the previous account field's change + whatever i updated in the following PR

**Reply by scolladon** [reply]
From what branch (or commit) the next PR is created ?
When the PR1 is merged, origin/master have the new changes.
Does the PR2 is created from an updated version of the master locally ? (Does a git pull on the master branch has been done after PR1 is merged on main and before creating the PR2 branch?)

**Reply by j-obrien97** [reply]
Yes, before making new changes i am doing a git pull on the origin/master branch and creating a new feature branch off of that updated local copy. Additionally I have been checking the PR's modified files and only the files i expect to show up are there. No PR's contain the same 2 changes but still are producing this additive delta

**Reply by scolladon** [reply]
What is the result when you execute the plug-in locally on the PR2 branch ?

I do not reproduce on my side. Only the changes from my branch are present when executing the plugin, could you setup the repo so I can reproduce it please ?

**Reply by j-obrien97** [reply]
Its working when done locally on PR2 but my SGD commands do look a little different from what im using in the pipeline.
Locally i am using the "HEAD" format:
sfdx sgd:source:delta --to "HEAD" --from "HEAD~1" --output "."

However, that doesnt work when im in AzureDevOps, so i had to change the command to:
sfdx sgd:source:delta -f origin/master --output .

Would the omission of the -t be causing this issue

**Reply by scolladon** [reply]
I don't think the omission of the `-t` parameter change anything here as it is replaced by "HEAD" anyway.
What is the result of `sfdx sgd:source:delta -f origin/master --output .` locally ?

Do you have a way to share this repository ? Or a similar one just so I can reproduce it myself ?

**Reply by j-obrien97** [reply]
Locally its working as expected. 
(Local process)
Git Pull on Master branch
Created new branch 
updated a profile permission save & commit
Run SGD 
Create package with singular profile update and doesn't include 3 other profile updates + 1 class update

Unfortunately, i don't have the access to share the repo or a similar one. Its pretty bare bones though, only contains some Lightning Web Component (singular JSON), profiles, and a handful of classes. Its possible that AzureDevOps is doing some manipulating behind the scenes though. I tried adding -t "HEAD" -f "HEAD~1" as well and the HEAD~1 is resolving correctly

**Reply by scolladon** [reply]
If you can have a repo to isolate and reproduce it would be perfect.
It is not clear to me what is happening in the context of this pipeline and I need a closer look.

**Reply by j-obrien97** [reply, marked as answer]
Hey @scolladon 
Ive actually have been able to resolve my issues. 
So for the "HEAD~1" format not working correctly, I needed to set my fetchDepth to 0 in the steps piece of the yaml, that corrected my post merge pipeline. I wasnt aware but in 2019(?) Azure made it so that new pipelines fetchDepth was automatically set to shallow, so this piece is a necessity :
steps:
- checkout: self
  persistCredentials: true
  fetchDepth: 0

For the bringing back multiple changes into the delta. I found that using $(git rev-parse origin/master) is the way to go. I had been messing around with the commits but when i would try to add remotes/origin or just origin/ it was failing. But the command by itself seems to be working as expected

**Reply by scolladon** [reply]
Hi @j-obrien97 thanks for sharing this.

You are not the [first one](https://github.com/scolladon/sfdx-git-delta/issues/573#issuecomment-1514255018) to have this issue with Azure...
We have a dedicated [section](https://github.com/scolladon/sfdx-git-delta#cicd-specificity) about CI/CD, but it is not specific... Maybe we could have a different file just to cover the specificities of every CI/CD used with Salesforce...
